cpp/src/reductions/reduction_operators.cuh

```diff
@@ -1,58 +1,233 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CUDF_REDUCTION_OPERATORS_CUH
+#define CUDF_REDUCTION_OPERATORS_CUH
+
 #include <cudf/cudf.h>
 #include <utilities/cudf_utils.h>
 #include <utilities/wrapper_types.hpp>
 #include <utilities/error_utils.hpp>
+#include <iterator/iterator.cuh>
 
-#include <utilities/device_atomics.cuh>
+#include <utilities/device_operators.cuh>
+
+#include <cmath>
 
 namespace cudf {
-namespace reductions {
+namespace reduction {
 
-// ------------------------------------------------------------------------
-// difinitions of device struct for binary operation
+// intermediate data structure to compute `var`, `std`
+template<typename ResultType>
+struct var_std
+{
+    ResultType value;                /// the value
+    ResultType value_squared;        /// the value of squared
 
-struct IdentityLoader {
-    template<typename T>
-    __device__
-        T operator() (const T *ptr, int pos) const {
-        return ptr[pos];
-    }
+    CUDA_HOST_DEVICE_CALLABLE
+    var_std(ResultType _value=0, ResultType _value_squared=0)
+    : value(_value), value_squared(_value_squared)
+    {};
+
+    using this_t = cudf::reduction::var_std<ResultType>;
+
+    CUDA_HOST_DEVICE_CALLABLE
+    this_t operator+(this_t const &rhs) const
+    {
+        return this_t(
+            (this->value + rhs.value),
+            (this->value_squared + rhs.value_squared)
+        );
+    };
+
+    CUDA_HOST_DEVICE_CALLABLE
+    bool operator==(this_t const &rhs) const
+    {
+        return (
+            (this->value == rhs.value) &&
+            (this->value_squared == rhs.value_squared)
+        );
+    };
 };
 
-struct SquaredLoader {
-    template<typename T>
-    __device__
-    T operator() (const T* ptr, int pos) const {
-        T val = ptr[pos];   // load
-        return val * val;   // squared
-    }
+// transformer for `struct var_std` in order to compute `var`, `std`
+template<typename ResultType>
+struct transformer_var_std
+{
+    using OutputType = cudf::reduction::var_std<ResultType>;
+
+    CUDA_HOST_DEVICE_CALLABLE
+    OutputType operator() (ResultType const & value)
+    {
+        return OutputType(value, value*value);
+    };
 };
 
-struct ReductionSum {
-    using Loader = IdentityLoader;
+// ------------------------------------------------------------------------
+// difinitions of device struct for reduction operation
+// all `op::xxx` must have `Op` and `make_iterator`
+// `Op`  is used to compute the reduction at device
+// `make_iterator` is used to make a iterator to be used to compute the reduction at device
+
+namespace op {
+
+// `product`, `product`, `sum_of_squares`, `min`, `max`
+// are used at simple_reduction
+
+struct sum {
     using Op = cudf::DeviceSum;
+
+    template<bool has_nulls, typename ElementType, typename ResultType>
+    static auto make_iterator(gdf_column const& column, ResultType identity)
```

Comment:
`identity` doesn't need to be an argument for any of these operators. Since `make_iterator` is a member, it already knows what its `identity` value is.

```
static auto make_iterator(gdf_column const& column){
        return cudf::make_iterator<has_nulls, ElementType, ResultType>(column, Op::template identity<ResultType>);
}
```

Reply:
Exactly.

Reply:
Updated. Done